### ISSUE TITLE
Gradual increase in memory usage by set_map_to_strage_cb

### ISSUE THREAD

**Issue by kjunichi** [trusterd/mruby-http2#25]
I tested on this environment.

```bash
ProductName:    Mac OS X
ProductVersion: 10.10.2
BuildVersion:   14C1510
```
and I used mruby-http2 in this commit.

```bash
git log|head
commit 16aa812fb4a600ae87f0924c0e425419a2e5d83b
Merge: 5bd8800 c41b8f7
Author: MATSUMOTO, Ryosuke <matsumoto_r@net.ist.i.kyoto-u.ac.jp>
Date:   Tue Mar 10 13:02:01 2015 +0900

    Merge pull request #24 from trusterd/refactor-client
    
    Add HTTP2::Clietn#{create_session,delete_session,request_stream}

commit c41b8f727d5f0a22db068ec6e6366ec4a089cd18
```

 

test.conf.rb is 

- https://gist.github.com/kjunichi/e5af76f144e0dc065d58

then, I ran the test at these commands.

```bash
./mruby/bin/mruby test.conf.rb &
./mruby/build/host/mrbgems/mruby-http2/nghttp2/src/h2load -n1000000 http://127.0.0.1:8080/mem
```

the result is this.

```
27067  mruby        0.0   00:40.04 1    0    15    188M   0B     0B     27067 26815 sleeping *0[1]           0.00000
```

thank you.

**Comment by matsumoto-r** [comment]
@kjunichi thank you for your report.

I checked this issue, but can't reproduce this problem using below configuration.

```ruby
s.set_map_to_strage_cb {
  s.request.uri
}
```

Could you reproduce this issue using more simple configuration?

**Comment by kjunichi** [comment]
I wrote more simple configuration.

```rb
SERVER_NAME = "Trusterd"
SERVER_VERSION = "0.0.1"
SERVER_DESCRIPTION = "#{SERVER_NAME}/#{SERVER_VERSION}"

root_dir = "."

s = HTTP2::Server.new({
  :port           => 8080,
  :document_root  => "#{root_dir}/htdocs",
  :server_name    => SERVER_DESCRIPTION,
  :worker         => "auto",
  :tls => false,
  :callback => true,
})

s.set_map_to_strage_cb {
  s.request.uri
}
s.run
```

I tried to not url where not exist in under htdocs

```
mruby/build/host/mrbgems/mruby-http2/nghttp2/src/h2load -n1000000 http://127.0.0.1:8080/test
```

but I try exits url like index.html,It's no increase memory usage.

```bash
mruby/build/host/mrbgems/mruby-http2/nghttp2/src/h2load -n1000000 http://127.0.0.1:8080/index.html
```

and

It was not seen on Linux.
It seems OSX specific problem at this time.

**Comment by kjunichi** [comment]
Thanks for your advice about valgrind.

I added some TRACER; statements.

I found that three is no delete operation at http2_session_data->nva.
And nghttp2_session_get_stream_user_data returns not collect http2_session_data->nvlen.

```
    >>>> mrb_http2_server_init:2300
    >>>> mrb_http2_worker_run:2080
    >>>> mrb_start_listen:2011
    >>>> mrb_start_listen:2024
warning: don't set run_user, so run with uid=501
    >>>> mrb_http2_acceptcb:1846
    >>>> create_http2_session_data:1698
    >>>> create_http2_session_data:1721
    >>>> mrb_http2_server_session_init:1584
    >>>> send_server_connection_header:1623
    >>>> send_server_connection_header:1628
    >>>> mrb_http2_server_readcb:1760
    >>>> session_recv:299
    >>>> server_on_frame_recv_callback:1488
    >>>> server_on_begin_headers_callback:1154
    >>>> create_http2_stream_data:161
    >>>> add_stream:138
    >>>> server_on_begin_headers_callback:1164
    >>>> percent_decode:364
    >>>> percent_decode:383
    >>>> server_on_frame_recv_callback:1488
    >>>> mrb_http2_process_request:1318
    >>>> mrb_http2_process_request:1330
    >>>> callback_ruby_block:113
    >>>> mrb_http2_server_set_content_cb:2217
    >>>> callback_ruby_block:116
    >>>> content_cb_reply:843
    >>>> callback_ruby_block:113
    >>>> callback_ruby_block:116
    >>>> callback_ruby_block:118
    >>>> fixup_status_header:1180
>>CREATE<<[:status]
>>r->reshdrslen[1]<<
>>CREATE<<[server]
>>CREATE<<[date]
    >>>> content_cb_reply:886
>>CREATE<<[content-length]
    >>>> send_response:502
>>nvlen = [4]
    >>>> mrb_http2_request_rec_free:6
    >>>> send_response:521
    >>>> content_cb_reply:906
    >>>> session_recv:309
    >>>> session_send:276
    >>>> server_send_callback:324
    >>>> server_send_callback:336
    >>>> server_send_callback:324
    >>>> server_send_callback:336
    >>>> server_send_callback:324
    >>>> server_send_callback:336
    >>>> file_read_callback:467
    >>>> file_read_callback:480
    >>>> server_send_callback:324
    >>>> server_send_callback:336
    >>>> server_on_stream_close_callback:1559
>>nvlen = [0]
    >>>> remove_stream:148
    >>>> delete_http2_stream_data:190
>>DELETE nvlen<<[0]
    >>>> server_on_stream_close_callback:1567
```

So, Unfortunatly this code is not delete the nva items.

```
static void delete_http2_stream_data(mrb_state *mrb,
    http2_session_data *session_data, http2_stream_data *stream_data)
{
  TRACER;
  if(stream_data->fd != -1) {
    close(stream_data->fd);
  } else {
	  fprintf(stderr,"%s","stream_data->fd = -1!");
  }
  mrb_free(mrb, stream_data->unparsed_uri);
  mrb_free_unless_null(mrb, stream_data->percent_encode_uri);
  if (stream_data->request_args != NULL) {
    mrb_free(mrb, stream_data->request_path);
    mrb_free(mrb, stream_data->request_args);
  }
  if (stream_data->request_body != NULL) {
    mrb_free(mrb, stream_data->request_body);
  }
  if (stream_data->upstream_req != NULL) {
    evhttp_request_free(stream_data->upstream_req);
  }
  if (session_data->app_ctx->server->config->server_status) {
    session_data->app_ctx->server->worker->active_stream--;
  }
  printf(">>DELETE nvlen<<[%d]\n", stream_data->nvlen);
  if (stream_data->nvlen > 0) {
    int i = stream_data->nvlen;
    for(;i>0;i--) {
      if(stream_data->nva[i-1].name != NULL) {
        printf(">>DELETE<<[%s]\n",stream_data->nva[i-1].name);
        mrb_free(mrb,stream_data->nva[i-1].name);
        stream_data->nva[i-1].name = NULL;
      }
      if(stream_data->nva[i-1].value != NULL) {
        mrb_free(mrb,stream_data->nva[i-1].value);
        stream_data->nva[i-1].value = NULL;
      }
    }

  }
  mrb_free(mrb, stream_data);
}
```

**Comment by matsumoto-r** [comment]
@kjunichi stream_data->nva and nvlen were freed by mrb_http2_free_nva() in mrb_http2_request_rec_free().

**Comment by matsumoto-r** [comment]
I don't investigate, but r->reshdrs was not freed after submitting response. This issue may be caused by the r->reshdrs.

**Comment by kjunichi** [comment]
I rewrote debug codes.

i got result about this problem.

these header items seems to be OK.

```
>>CREATE<<[accept]
>>stream_data->nvlen[1]
>>CREATE<<[accept-encoding]
>>stream_data->nvlen[2]
>>CREATE<<[user-agent]
>>stream_data->nvlen[3]
```

```
>>DELETE[accept]
>>DELETE[accept-encoding]
>>DELETE[user-agent<]
```

but these header items are not.

```
>>CREATE<<[:status]
>>r->reshdrslen[1]<<
>>CREATE<<[server]
>>CREATE<<[date]
>>>> content_cb_reply:888
>>CREATE<<[content-length]
```

these items ware not freed.

I wonder why valgrind said OK on linux...

all result is here.

```
mruby/bin/mruby simple.conf.rb
>>>> mrb_http2_server_init:2302
>>>> mrb_http2_worker_run:2082
>>>> mrb_start_listen:2013
>>>> mrb_start_listen:2026
warning: don't set run_user, so run with uid=501
>>>> mrb_http2_acceptcb:1848
>>>> create_http2_session_data:1700
>>>> create_http2_session_data:1723
>>>> mrb_http2_server_session_init:1586
>>>> send_server_connection_header:1625
>>>> send_server_connection_header:1630
>>>> mrb_http2_server_readcb:1762
>>>> session_recv:300
>>>> server_on_frame_recv_callback:1490
>>>> server_on_begin_headers_callback:1156
>>>> create_http2_stream_data:161
>>>> add_stream:138
>>>> server_on_begin_headers_callback:1166
>>>> percent_decode:365
>>>> percent_decode:384
>>CREATE<<[accept]
>>stream_data->nvlen[1]
>>CREATE<<[accept-encoding]
>>stream_data->nvlen[2]
>>CREATE<<[user-agent]
>>stream_data->nvlen[3]
>>>> server_on_frame_recv_callback:1490
>>>> mrb_http2_process_request:1320
>>>> mrb_http2_process_request:1332
>>>> callback_ruby_block:113
>>>> mrb_http2_server_set_content_cb:2219
>>>> callback_ruby_block:116
>>>> content_cb_reply:845
>>>> callback_ruby_block:113
>>>> callback_ruby_block:116
>>>> callback_ruby_block:118
>>>> fixup_status_header:1182
>>CREATE<<[:status]
>>r->reshdrslen[1]<<
>>CREATE<<[server]
>>CREATE<<[date]
>>>> content_cb_reply:888
>>CREATE<<[content-length]
>>>> send_response:503
>>nvlen = [4]
>>>> send_response:520
>>r->reqhdrlen = [3]
>>>> mrb_http2_request_rec_free:6
>>>> mrb_http2_request_rec_free:31
>>>> mrb_http2_free_nva:138
>>nvlen = [3]
>>DELETE[accept]
>>DELETE[accept-encoding]
>>DELETE[user-agent<]
>>>> send_response:523
>>>> content_cb_reply:908
>>>> session_recv:310
>>>> session_send:277
>>>> server_send_callback:325
>>>> server_send_callback:337
>>>> server_send_callback:325
>>>> server_send_callback:337
>>>> server_send_callback:325
>>>> server_send_callback:337
>>>> file_read_callback:468
>>>> file_read_callback:481
>>>> server_send_callback:325
>>>> server_send_callback:337
>>>> server_on_stream_close_callback:1561
>>nvlen = [3]
>>>> remove_stream:148
>>>> delete_http2_stream_data:190
>>>> server_on_stream_close_callback:1569
>>>> session_send:283
>>>> session_recv:314
>>>> mrb_http2_server_writecb:1782
>>>> mrb_http2_server_writecb:1782
>>>> mrb_http2_server_writecb:1786
>>>> mrb_http2_server_writecb:1792
>>>> session_send:277
>>>> session_send:283
>>>> mrb_http2_server_writecb:1804
>>>> mrb_http2_server_readcb:1762
>>>> session_recv:300
>>>> server_on_frame_recv_callback:1490
>>>> server_on_frame_recv_callback:1490
>>>> session_recv:310
>>>> session_send:277
>>>> session_send:283
>>>> session_recv:314
>>>> mrb_http2_server_eventcb:1814
>>>> mrb_http2_server_eventcb:1837
>>>> delete_http2_session_data:239
>>>> mrb_http2_conn_rec_free:128
>>>> mrb_http2_acceptcb:1848
>>>> create_http2_session_data:1700
>>>> create_http2_session_data:1723
>>>> mrb_http2_server_session_init:1586
>>>> send_server_connection_header:1625
>>>> send_server_connection_header:1630
>>>> mrb_http2_server_writecb:1782
>>>> mrb_http2_server_writecb:1786
>>>> mrb_http2_server_writecb:1792
>>>> session_send:277
>>>> server_send_callback:325
>>>> server_send_callback:337
>>>> session_send:283
>>>> mrb_http2_server_writecb:1804
>>>> mrb_http2_server_readcb:1762
>>>> session_recv:300
>>>> session_recv:310
>>>> session_send:277
>>>> session_send:283
>>>> session_recv:314
>>>> mrb_http2_server_writecb:1782
>>>> mrb_http2_server_writecb:1786
>>>> mrb_http2_server_writecb:1792
>>>> session_send:277
>>>> session_send:283
>>>> mrb_http2_server_writecb:1804
>>>> mrb_http2_server_readcb:1762
>>>> session_recv:300
>>>> server_on_frame_recv_callback:1490
>>>> server_on_begin_headers_callback:1156
>>>> create_http2_stream_data:161
>>>> add_stream:138
>>>> server_on_begin_headers_callback:1166
>>>> percent_decode:365
>>>> percent_decode:384
>>CREATE<<[accept]
>>stream_data->nvlen[1]
>>CREATE<<[accept-encoding]
>>stream_data->nvlen[2]
>>CREATE<<[user-agent]
>>stream_data->nvlen[3]
>>>> server_on_frame_recv_callback:1490
>>>> mrb_http2_process_request:1320
>>>> mrb_http2_process_request:1332
>>>> callback_ruby_block:113 
```

**Comment by matsumoto-r** [comment]
@kjunichi thank you for your investigation. Could you create the pull-request which fix this memory leak on your OSX? I'll also check your code on my linux.

**Comment by matsumoto-r** [comment]
Fixed via #30
